### PR TITLE
Issue #408: Handling of preference header 'continue-on-error' in user…

### PR DESCRIPTION
**Pull request by TomDu**
Handling of preference header 'continue-on-error' in user request according to the protocol spec.

### CODE REVIEW

File: OData/src/System.Web.OData/OData/Batch/ChangeSetResponseItem.cs [hunk 1]
```diff
@@ -56,6 +56,21 @@ public ChangeSetResponseItem(IEnumerable<HttpResponseMessage> responses)
             writer.WriteEndChangeset();
         }
 
+        /// <summary>
+        /// Gets a value that indicates if the HTTP response in response item is successful.
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Batch/ChangeSetResponseItem.cs]: if the responses in this item are successful.

File: OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs [hunk 2]
```diff
@@ -43,6 +43,11 @@ public ODataMessageQuotas MessageQuotas
         public string ODataRouteName { get; set; }
 
         /// <summary>
+        /// Gets or sets the value That defines whether the batch request contains preference 'continue-on-error'.
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs]: That -> that

File: OData/src/System.Web.OData/OData/Batch/ODataBatchResponseItem.cs [hunk 3]
```diff
@@ -88,6 +88,14 @@ public void Dispose()
         public abstract Task WriteResponseAsync(ODataBatchWriter writer, CancellationToken cancellationToken);
 
         /// <summary>
+        /// Gets a value that indicates if the HTTP response in response item is successful.
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Batch/ODataBatchResponseItem.cs]: if the responses in this item are successful.

File: OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs [hunk 4]
```diff
@@ -48,6 +48,14 @@ public override Task WriteResponseAsync(ODataBatchWriter writer, CancellationTok
             return WriteMessageAsync(writer, Response, cancellationToken);
         }
 
+        /// <summary>
+        /// Gets a value that indicates if the HTTP response in response item was successful.
+        /// </summary>
+        internal override bool IsResponseSuccessful()
```
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs]: Is this method necessary?

File: OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs [hunk 5]
```diff
@@ -473,4 +473,161 @@ public static void UpdateWebConfig(WebConfigHelper webConfig)
...
+                GetEdmModel(),
+                new DefaultODataPathHandler(),
+                ODataRoutingConventions.CreateDefault(),
+                new DefaultODataBatchHandler(server));
+        }
+
+        protected static IEdmModel GetEdmModel()
+        {
+            ODataModelBuilder builder = new ODataConventionModelBuilder();
+            EntitySetConfiguration<DefaultBatchCustomer> customers = builder.EntitySet<DefaultBatchCustomer>("DefaultBatchCustomer");
```
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: Better to name the entity set as "XXXs"

File: OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs [hunk 6]
```diff
@@ -473,4 +473,161 @@ public static void UpdateWebConfig(WebConfigHelper webConfig)
...
+                        case ODataBatchReaderState.Operation:
+                            var operationMessage = batchReader.CreateOperationResponseMessage();
+                            subResponseCount++;
+                            break;
+                    }
+                }
+            }
+
+            // Assert
+            Assert.Equal(1, subResponseCount);
```
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: Need to test the failure?

File: OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs [hunk 7]
```diff
@@ -473,4 +473,161 @@ public static void UpdateWebConfig(WebConfigHelper webConfig)
...
+                        case ODataBatchReaderState.Operation:
+                            var operationMessage = batchReader.CreateOperationResponseMessage();
+                            subResponseCount++;
+                            break;
+                    }
+                }
+            }
+
+            // Assert
+            Assert.Equal(3, subResponseCount);
```
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/Batch/Tests/DataServicesClient/DefaultODataBatchHandlerTests.cs]: at least to test the status code of each response

File: OData/test/UnitTest/System.Web.OData.Test/OData/Batch/ChangeSetResponseItemTest.cs [hunk 8]
```diff
@@ -69,6 +69,30 @@ public void WriteResponseAsync_WritesChangeSet()
         }
 
         [Fact]
+        public void IsResponseSuccessful_TestResponse()
+        {
+            // Arrange
```
Comment (xuzhg) [inline OData/test/UnitTest/System.Web.OData.Test/OData/Batch/ChangeSetResponseItemTest.cs]: // Arrange 

// Act

// Assert

File: OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs (hunk 4 above)
Comment (TomDu) [inline OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs]: @xuzhg It's an override method that overrides the base member, I did this in order to keep consistency of the interface.

**Comment by xuzhg** [comment]
Merged. Thanks

File: OData/src/System.Web.OData/OData/Batch/DefaultODataBatchHandler.cs [hunk 9]
```diff
@@ -41,6 +41,15 @@ public DefaultODataBatchHandler(HttpServer httpServer)
...
+            string preferHeader = RequestPreferenceHelpers.GetRequestPreferHeader(request);
+            if (preferHeader != null && preferHeader.Contains("odata.continue-on-error"))
+            {
+                ContinueOnError = true;
+            }
+            else
+            {
+                ContinueOnError = false;
+            }
             try
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/DefaultODataBatchHandler.cs]: Add a whitespace.

File: OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs [hunk 10]
```diff
@@ -48,6 +48,14 @@ public override Task WriteResponseAsync(ODataBatchWriter writer, CancellationTok
             return WriteMessageAsync(writer, Response, cancellationToken);
         }
 
+        /// <summary>
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/OperationResponseItem.cs]: Comments are optional for internal, private methods.

File: OData/src/System.Web.OData/OData/Batch/DefaultODataBatchHandler.cs [hunk 11]
```diff
@@ -41,6 +41,15 @@ public DefaultODataBatchHandler(HttpServer httpServer)
 
             IList<ODataBatchRequestItem> subRequests = await ParseBatchRequestsAsync(request, cancellationToken);
 
+            string preferHeader = RequestPreferenceHelpers.GetRequestPreferHeader(request);
+            if (preferHeader != null && preferHeader.Contains("odata.continue-on-error"))
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/DefaultODataBatchHandler.cs]: const

File: OData/src/System.Web.OData/OData/Batch/UnbufferedODataBatchHandler.cs [hunk 12]
```diff
@@ -49,6 +49,16 @@ public UnbufferedODataBatchHandler(HttpServer httpServer)
             List<ODataBatchResponseItem> responses = new List<ODataBatchResponseItem>();
             Guid batchId = Guid.NewGuid();
             List<IDisposable> resourcesToDispose = new List<IDisposable>();
+
+            string preferHeader = RequestPreferenceHelpers.GetRequestPreferHeader(request);
+            if (preferHeader != null && preferHeader.Contains("odata.continue-on-error"))
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/UnbufferedODataBatchHandler.cs]: const

File: OData/src/System.Web.OData/OData/Batch/ChangeSetResponseItem.cs [hunk 13]
```diff
@@ -56,6 +56,21 @@ public ChangeSetResponseItem(IEnumerable<HttpResponseMessage> responses)
             writer.WriteEndChangeset();
         }
 
+        /// <summary>
+        /// Gets a value that indicates if the responses in this item are successful.
+        /// </summary>
+        internal override bool IsResponseSuccessful()
+        {
+            foreach (HttpResponseMessage response in Responses)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/ChangeSetResponseItem.cs]: `return Responses.All(r => r.IsSuccessStatusCode);`

File: OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs [hunk 14]
```diff
@@ -43,6 +43,11 @@ public ODataMessageQuotas MessageQuotas
         public string ODataRouteName { get; set; }
 
         /// <summary>
+        /// Gets or sets the value that defines whether the batch request contains preference 'continue-on-error'.
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs]: Add this public API change to a doc or page.

File: OData/test/UnitTest/System.Web.OData.Test/OData/Batch/UnbufferedODataBatchHandlerTest.cs [hunk 15]
```diff
@@ -168,6 +168,81 @@ public void ProcessBatchAsync_DisposesResponseInCaseOfException()
...
+                    ODataBatchRequestHelper.CreateODataRequestContent(new HttpRequestMessage(HttpMethod.Post, "http://example.com/values")
+                    {
+                        Content = new StringContent("bar")
+                    }),
+                }
+            };
+            batchRequestWithPrefContinueOnError.Headers.Add("prefer", "odata.continue-on-error");
+
+            // Act
+            var response = batchHandler.ProcessBatchAsync(batchRequest, CancellationToken.None).Result;
```
Comment (congysu) [inline OData/test/UnitTest/System.Web.OData.Test/OData/Batch/UnbufferedODataBatchHandlerTest.cs]: Check the batch response has a `Preference-Applied` header with values include `odata.continue-on-error`.

File: OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs (hunk 14 above)
Comment (chgarg) [inline OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs]: consider renaming ContinueOnError to IsPreferenceHeaderContinueOnError

File: OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs [hunk 16]
```diff
@@ -43,6 +43,11 @@ public ODataMessageQuotas MessageQuotas
...
+        /// </summary>
+        public bool ContinueOnError { get; set; }
```
Comment (chgarg) [inline OData/src/System.Web.OData/OData/Batch/ODataBatchHandler.cs]: consider renaming ContinueOnError to IsPreferenceHeaderContinueOnError

**Pull request closed without merging**